### DISCUSSION TITLE
Prove generic type conditions

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#6731] [Language Ideas] [answered]
Today the following code does not compile

``` csharp
using System;
using System.Collections.Generic;

static class PropertyWrapper
{

    static readonly Dictionary<string, object> Properties = new();

    static T GetProperty<T>(string key) => Properties[key] switch
    {
        T obj => obj,
        string v when typeof(T).IsEnum && typeof(T).IsValueType && Enum.TryParse<T>(v, out T? parsed) => parsed,
        var ob => (T)ob
    };
}
```

with the error `error CS0453: The type 'T' must be a non-nullable value type in order to use it as parameter 'TEnum' in the generic type or method 'Enum.TryParse<TEnum>(ReadOnlySpan<char>, out TEnum)'`

However, we can exhaustively prove that `T` is, in fact, a non-nullable value type. The compiler should be able to recognize this.

**Comment by HaloFour** [comment, marked as answer]
I think this is applicable: #6308
